### ISSUE TITLE
XAML type provider

### ISSUE THREAD

**Issue by bender2k14** [elmish/Elmish.WPF#142]
This issue spun out of https://github.com/elmish/Elmish.WPF/issues/141#issuecomment-538105026.

The general question is
> Is it possible to improve type safety by using XAML type provider?

**Comment by bender2k14** [comment]
Replying to https://github.com/elmish/Elmish.WPF/issues/141#issuecomment-538117978.

> If you introduce a type provider in Elmish.WPF, then perhaps consider make it optional in some way if it isn't too much of a bother. E.g. make it possible to bind with either magic string or through the type provider, so that if the type provider fails, at least the magic string mechanism saves the day.

Yes, that is my thinking as well.  The magic string approach isn't that bad, especially when you consider that essentially all GUI frameworks do something similar.  I don't have any current plans to look into this, but I hope to do so eventually.  Thanks for the links and short descriptions :)

**Comment by BentTranberg** [comment]
I agree. Magic strings definitely isn't the biggest problem we deal with. Nevertheless GUI breaks from time to time because a string depended on some identifier that was refactored (renamed). Using dedicated local types for bindings and streamings, or running unit tests that checks with 'nameof', are examples of how to avoid or detect such breaking changes. I like doing heavy refactorings to keep my code tidy, so I also depend heavily on that kind of security around my code. It works.

**Comment by cmeeren** [comment]
If you use design-time VMs (check the readme), the tooling might warn you if there's a binding from XAML that don't exist in the VM.

Not sure if design-time VMs work with netcoreapp3.0 by the way, because it seems that the XAML designer currently doesn't support Project Code.

**Comment by bender2k14** [comment]
> Nevertheless GUI breaks from time to time because a string depended on some identifier that was refactored (renamed). ... I like doing heavy refactorings to keep my code tidy, so I also depend heavily on that kind of security around my code. It works.

Me too.  My goal is to have all the magic strings be string literals passed into Elmish.WPF bindings.  That way, I can't break anything with using Visual Studio's Refactor-Rename.  After I make the changed described in https://github.com/elmish/Elmish.WPF/issues/131#issuecomment-538398038, I will once again be achieving this goal with my work project.

**Comment by bender2k14** [comment]
> ...it seems that the XAML designer currently doesn't support Project Code.

What is "Project Code"?

**Comment by cmeeren** [comment]
I'm no expert, but the designer basically runs your project so that you get the design-time data you want.

**Comment by BentTranberg** [comment]
> If you use design-time VMs

I do, but I don't reuse the runtime model for that. I find it much easier to simply create a class or two or three with properties and whatnots. My goal, obviously, is just to deliver some fake data for design time. I couldn't care less whether the runtime and design-time VM's are properly aligned in all manners. After all, there is no interactivity either, so why bother with a more complex model that includes the time aspect when the data is all static. The visual design generally becomes easier this way, at least for me.

**Comment by cmeeren** [comment]
Some benefits of a using runtime model are 1) you're guaranteed to keep the design-VM props in sync with the bindings (since they're the same), and 2) you already have the bindings, so you don't need to create another class.

Making a simple design-time VM It's a perfectly reasonable tradeoff, though.

**Comment by bender2k14** [comment]
This issue is related to #13

**Comment by bender2k14** [comment]
In case its helpful, Pluralsight has a course about how to create a type provider.
[Building F# Type Providers by Dave Fancher](https://www.pluralsight.com/courses/building-f-sharp-type-providers)

**Comment by light-pigeon** [comment]
@cmeeren said:
> ...giving you useful Intellisense and warnings when changing/renaming the bindings in `bindings`

But I cannot see a warning in the designer when the binding does not match the binding name on the VM. (see animation on sample project)

![Animation](https://user-images.githubusercontent.com/42803583/99081893-7fa8de80-25d4-11eb-9308-6e6fb39b877c.gif)

**Comment by bent-rasmussen** [comment]
@light-pigeon  Try this instead:

```xml
d:DataContext="{d:DesignInstance Type=vm:AppWorkspaceViewModel, IsDesignTimeCreatable=False}"
```

It works with Intellisense and gives a warning if the identifier is misspelt (using .NET 5 and VS 16.8.3) (of course you can probably omit the IsDesignTimeCreatable).

Currently looking at Elmish.WPF as a way to modernize our fully F# WPF app (yes, that was a stupid decision looking at tooling) but the lack of static binding - when looking at the samples - is a step backwards compared to our current oldschool approach; we love static and design-time typing. ;-)

Using a type provider would be great but FsXaml.Wpf looks like abandonware - the Nuget package was last published close to 4 years ago and the last repository commit was in 2018.

**Comment by TysonMN** [comment]
I useful link it this issue is to the [FsXaml project](https://github.com/fsprojects/FsXaml) (which was lasted updated in 2018).

**Comment by marner2** [comment]
For the record, I have a solution to this documented in #494. It involves replacing the type lists into an F# ViewModel type, implemented through a helper. This solution therefore leverages the types directly. It should work completely inline with the current solution.

**Comment by TysonMN** [comment]
Very interesting. I am excited to try this.
